### ISSUE TITLE
OOM due to large thead_table

### ISSUE THREAD

**Issue by timohochberger** [DynamoRIO/dynamorio#3401]
Hi,

we are facing a bug with dynamorio and hoping to get some help with it. Please find a detailed description below.

* Setup

Linux operating system on intel 64 bit platform.

* Introduction

Dynamorio is used with our application for some years now. The use case is to collect coverage information using drcov.

* Bug

The problem we are faced with is that the drrun process is occasionally aborted due to an out-of-memory situation (from
now on just referred to by OOM). 

* Reproduction scenario

To narrow down the problem we started reproduction with a reduced and artificial example in order to reduce complexity -
as reproduction in our setting is very hard to achieve. We have a code generation script synthesizing c++ code that
produces very large executables when being built. With this approach we were able to reproduce the OOM reliably. The
generated C++ code itself is very simple as it consists of a very high number of functions that are called sequentially.

With ~2 million functions split across ~100 compilation units we end up having a binary that is slightly larger in size
than 1 gigabyte. Executing drrun with such a binary always results in an OOM - even without any client.

* Steps to reproduce

Please find attached an archive containing the reproduction environment that basically consists of 1) a code generator
'generateCode.py' and 2) a wrapper script 'reproduceBug.sh' to simplify usage of the reproduction.

All You need to do is to set the env variable DRRUN_ROOT to a valid dynamorio installation and run 'reproduceBug.sh'.

This will not only generate the code, but also build it and start the executable in order to provoke the bug.

All output can be found in a folder named 'generated-<timestamp>' in the current working directory.

If the error occurred, the line '<Out of memory.  Program aborted.>' should be printed to the shell.

* Expected results

The program runs through with rc 0 and no OOM.

* Actual results:

Many hash tables and caches of dynamorio grow constantly in size but are reset periodically by fcache_reset_all_caches_proactively. However, thead_table is not. When this table is resized from 128 MB to 256 MB, there is not enough contiguous memory left, and dynamorio aborts after reporting the OOM.

Clearing, freeing and recreating the thead_table of every thread in fcache_reset_all_caches_proactively makes the artificial example run through successfully. It has, however, a big performance impact for our production code.

Additionally, we are not 100% sure if the continuous size increase of thead_table is the root cause for the OOM situations we see with our production code as well. We have, however, many long running threads. As far as we understood, the thead_table is only cleared when a thread finishes execution, so it might be related.

Thanks a lot in advance!

Kind regards
Timo

[example.zip](https://github.com/DynamoRIO/dynamorio/files/2915518/example.zip)

**Comment by derekbruening** [comment]
Usually a 64-bit OOM is due to lack of 32-bit-displacement-reachable memory.    #774's feature to split vmheap and vmcode will help substantially by letting the heap go anywhere, but it's not there yet.

The runtime option `-vm_size 1G` may help (we should probably make this the default for 64-bit).

Shrinking the per-thread usage may also help: something like `-no_guard_pages -initial_heap_nonpers_size 12K -initial_heap_unit_size 4K -stack_size 24K -signal_stack_size 24K`.

Disabling traces would eliminate thead_table: `-disable_traces`.

**Comment by derekbruening** [comment]
thead_table is deliberately not cleared to avoid losing key profiling state.

The split vmheap and vmcode is #1132 (not #774 as referenced above) was just merged in PR #3554.  It will result in never hitting OOM on heap allocations simply due to reachability, thus improving scalability.  Does it solve the problem here?

**Comment by timohochberger** [comment]
We verified that the OOM situation does not occur anymore after #1132, so I'm closing this issue. We did run however into some kind of hanging situation now, but we will open a new issue for this. Thanks for your help!
